[Li Zhong Fu] ip

zhongfu opened the pull request:
# Duke

> Lorem ipsum dolor sit amet

It's a to-do list app, I guess.

- foo
- ~bar~
- **baz**

1. __qux__
2. *quux*
3. corge :tada: 

- [x] Why are we still here?
- [ ] Just to `suffer`?
- [ ] obligatory [link](https://youtu.be/dQw4w9WgXcQ)

```bash
# obligatory code block with syntax highlighting
:(){:|:&};:
```

zhongfu committed:
feat: Duke Level-1

zhongfu committed:
feat: Duke Level-2

zhongfu committed:
fix: use return (from `main()`) instead of System.exit to silence linter

zhongfu committed:
chore: add vscode config dir to gitignore

zhongfu committed:
refactor: use generics for enumerateList param

zhongfu committed:
feat: represent tasks as Task objects

zhongfu committed:
style: fix formatting

zhongfu committed:
refactor: add single-line dialog method

zhongfu committed:
feat: mark tasks as done or undone

zhongfu committed:
feat: add informative messages

zhongfu committed:
feat: users have to add tasks explicitly now

zhongfu committed:
feat: add new task types (deadline, event, todo)

zhongfu committed:
feat: add deadline/todo commands to chatbot

zhongfu committed:
test: add simple tests, update gitignore

zhongfu committed:
fix: ignore surrounding whitespace when parsing task arg strings

zhongfu committed:
feat: warn if task name or time is empty

zhongfu committed:
feat: add generic DukeException for... unexpected input?

zhongfu committed:
refactor: remove unreachable if-branch

zhongfu committed:
feat: add delete command

zhongfu committed:
tests: update text-ui-test

zhongfu committed:
fix: split lines in dialog(String) so that message will be indented properly

zhongfu committed:
refactor: types into new package

zhongfu committed:
feat: load/save tasks

zhongfu committed:
feat: use LocalDateTime instead of Strings for dates

zhongfu committed:
Merge branch 'branch-Level-7'
- feat: load/save tasks

zhongfu committed:
Merge branch 'branch-Level-8'
- feat: use LocalDateTime instead of Strings for dates

zhongfu committed:
fix: allow classes from java.time to be deserialized

zhongfu committed:
refactor: Storage as non-static class

zhongfu committed:
refactor: task list as non-static class

zhongfu committed:
refactor: chatbot task types -> chatbot.types.tasks

zhongfu committed:
fix: missing (+ extra) import in TaskList

zhongfu committed:
refactor: parser as separate class

zhongfu committed:
refactor: output stuff to separate class
i know it says UserInterface even though it's only used to print stuff out, i promise i'll fix it soon^tm

zhongfu committed:
refactor: Duke as non-static class

zhongfu committed:
refactor: move Parser into UserInterface

zhongfu committed:
refactor: move other ui-related classes into chatbot.types.ui

zhongfu committed:
fix(parser): strip positional args

zhongfu committed:
feat(types/tasks): add equals(), hashCode()

zhongfu committed:
fix(parser): make datetime resolver stricter
(e.g. reject invalid dates like 30 Feb 2021)

zhongfu committed:
tests: add tests for Storage and Parser

Gnoznaug commented on src/main/java/li/zhongfu/cs2103/chatbot/Duke.java:
```diff
@@ -0,0 +1,189 @@
+                            task = new ToDo(result.getPosArg());
+                        }
+                        
+                        this.tasks.add(task);
+                        this.ui.printDialog(new String[] {
+                            "New task added:",
+                            String.format(" %s", task),
+                            String.format("You now have %d tasks in your task list.", this.tasks.size())
+                        });
+                        break;
+
+                    case "reload":
+                        try {
+                            this.tasks = TaskList.loadFromStorage(this.storage);
+                            this.ui.printDialog(String.format("Tasks reloaded; %d tasks in list now", this.tasks.size()));
+                        } catch (FileNotFoundException e) {
+                            this.ui.printDialog("Saved tasks file not found! Skipping reload.");
+                        } catch (StorageException e) {
+                            this.ui.printDialog("Error loading tasks! Skipping reload.");
+                        }
+                        break;
+                    
+                    case "save":
+                        this.tasks.save(this.storage);
+                        this.ui.printDialog("Tasks saved!");
+                        break;
+
+                    case "bye":
+                        this.tasks.save(this.storage);
+                        this.ui.printDialog("Bye. Hope to see you again soon!");
+                        return;
+
+                    default:
+                        throw new DukeException("I don't know what that means! Try: todo, deadline, event, list, delete, mark, unmark, reload, save, bye");
+                }
+            } catch (DukeException e) {
+                this.ui.printDialog(e.getMessage().split("\n"));
+            }
+        }
+    }
```
According to the code quality guideline, the method is too long.
Source: https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/codeQuality.html

Gnoznaug commented on src/main/java/li/zhongfu/cs2103/chatbot/types/Storage.java:
```diff
@@ -0,0 +1,93 @@
+        this(filePath, TaskFilter::checkInput);
+    }
+
+    public List<Object> load() throws FileNotFoundException, StorageException {
+        List<Object> objs = new ArrayList<>();
+
+        try (FileInputStream fileStream = new FileInputStream(this.DATA_FILE);
+                ObjectInputStream objectStream = new ObjectInputStream(fileStream)) {
+            objectStream.setObjectInputFilter(this.OBJECT_FILTER);
+
+            while (true) {
+                Object o = objectStream.readObject();
+                if (o != null) {
+                    objs.add(o);
+                } else {
+                    logger.warning("Got null in objectStream, ignoring");
+                }
+            }
+        } catch (EOFException e) { // also means objectStream has no more objects, return normally
+            return objs;
+        } catch (FileNotFoundException e) { // throw FileNotFoundException instead of IOException/StorageException
+            throw e;
+        } catch (IOException e) {
+            throw new StorageException("Got IOException while trying to load objects", e);
+        } catch (ClassNotFoundException e) {
+            throw new StorageException("Unknown class found in storage!", e);
+        }
+    }
```
Code is easy to read and understand.

Gnoznaug commented on src/main/java/li/zhongfu/cs2103/chatbot/types/tasks/Task.java:
```diff
@@ -0,0 +1,76 @@
+ */
+public abstract class Task implements Serializable {
+    private String name;
+    private boolean done = false;
```
Maybe can follow the boolean naming convention.
```suggestion
    private boolean isDone = false;
```

Gnoznaug commented on src/main/java/li/zhongfu/cs2103/chatbot/types/ui/Parser.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Map;
+
+public class Parser {
+    private static final DateTimeFormatter DEFAULT_DATE_TIME_FORMATTER = new DateTimeFormatterBuilder()
+            .parseCaseInsensitive()
+            .appendPattern("[uuuu-MM-dd[ HH:mm[:ss]]['T'HH:mm[:ss]]]")
+            .appendPattern("[MMM d uuuu[ HH:mm[:ss]]]")
+            .appendPattern("[d MMM uuuu[ HH:mm[:ss]]]")
+            .appendPattern("[MMMM d uuuu[ HH:mm[:ss]]]")
+            .appendPattern("[d MMMM uuuu[ HH:mm[:ss]]]")
+            .parseDefaulting(ChronoField.HOUR_OF_DAY, 0)
+            .parseDefaulting(ChronoField.MINUTE_OF_HOUR, 0)
+            .parseDefaulting(ChronoField.SECOND_OF_MINUTE, 0)
+            .toFormatter()
+            .withResolverStyle(ResolverStyle.STRICT);
```
Very nice chaining makes it easy to read and understand.

Gnoznaug reviewed (commented):
I found your code easy to read for the most part.
I noted a few minor coding standard violations
too.
Overall, very nice!

zhongfu committed:
build: add gradle

zhongfu committed:
fix: add @Override decorator for toString()

zhongfu committed:
docs: add/improve javadocs

zhongfu committed:
style: add AB3 checkstyle config

zhongfu committed:
style: fix style to match Java Coding Standards

zhongfu committed:
Add find feature for searching tasks

zhongfu committed:
Add JUnit deps to Gradle, remove JUnit 4 usage

zhongfu committed:
Merge branch 'branch-A-JavaDoc'

zhongfu committed:
Merge branch 'branch-A-CodingStandard'

zhongfu committed:
Merge branch 'branch-Level-9'

zhongfu committed:
Add Checkstyle to Gradle

zhongfu committed:
Remove unused fields, imports

zhongfu committed:
Fix trailing whitespaces

wxliong commented on src/main/java/li/zhongfu/cs2103/chatbot/Duke.java:
```diff
@@ -0,0 +1,245 @@
+            } catch (DukeException e) {
+                this.ui.printDialog(e.getMessage().split("\n"));
+            }
+        }
```
Perhaps separating the functionality of each command into methods in the switch statement could help shorten it?

wxliong commented on src/main/java/li/zhongfu/cs2103/chatbot/types/ui/ParserResult.java:
```diff
@@ -0,0 +1,100 @@
+ */
+public class ParserResult {
+    private String cmd;
+    private Map<String, String> args;
```
Good usage of Map! I didn't think to use that for mine haha

wxliong commented on src/test/java/li/zhongfu/cs2103/chatbot/types/ui/ParserTest.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Test
+    public void parseDateTime_validFormats() {
+        // todo random years/months/days/etc
+        LocalDateTime expectedNoTime = LocalDateTime.of(2021, Month.NOVEMBER, 12, 0, 0);
+        LocalDateTime expectedNoSeconds = LocalDateTime.of(2021, Month.NOVEMBER, 12, 13, 14);
+        LocalDateTime expectedWithSeconds = LocalDateTime.of(2021, Month.NOVEMBER, 12, 13, 14, 15);
+
+        assertEquals(expectedNoTime, parser.parseDateTime("2021-11-12"));
+        assertEquals(expectedNoTime, parser.parseDateTime("nov 12 2021"));
+        assertEquals(expectedNoTime, parser.parseDateTime("November 12 2021"));
+        assertEquals(expectedNoTime, parser.parseDateTime("12 Nov 2021"));
+        assertEquals(expectedNoTime, parser.parseDateTime("12 november 2021"));
+        assertEquals(expectedNoSeconds, parser.parseDateTime("2021-11-12T13:14"));
+        assertEquals(expectedNoSeconds, parser.parseDateTime("2021-11-12 13:14"));
+        assertEquals(expectedNoSeconds, parser.parseDateTime("nov 12 2021 13:14"));
+        assertEquals(expectedNoSeconds, parser.parseDateTime("November 12 2021 13:14"));
+        assertEquals(expectedNoSeconds, parser.parseDateTime("12 Nov 2021 13:14"));
+        assertEquals(expectedNoSeconds, parser.parseDateTime("12 november 2021 13:14"));
+        assertEquals(expectedWithSeconds, parser.parseDateTime("2021-11-12T13:14:15"));
+        assertEquals(expectedWithSeconds, parser.parseDateTime("2021-11-12 13:14:15"));
+        assertEquals(expectedWithSeconds, parser.parseDateTime("nov 12 2021 13:14:15"));
+        assertEquals(expectedWithSeconds, parser.parseDateTime("November 12 2021 13:14:15"));
+        assertEquals(expectedWithSeconds, parser.parseDateTime("12 Nov 2021 13:14:15"));
+        assertEquals(expectedWithSeconds, parser.parseDateTime("12 november 2021 13:14:15"));
+    }
```
Great that you have many test cases to prevent issues that may be hard to spot!

wxliong commented on src/main/java/li/zhongfu/cs2103/chatbot/types/TaskList.java:
```diff
@@ -0,0 +1,160 @@
+        storage.save(this.list);
+    }
+
+    /**
+     * Returns a list of strings representing the tasks in this TaskList as
+     * an (1-indexed) ordered list.
+     *
+     * Each string in the list is of the form {@code n. foo}, where {@code n} is the
+     * index of the item (starting from 1)
+     * and {@code foo} is the string representation of the item.
+     *
+     * @return a list of strings representing the tasks in this TaskList as an 1-indexed ordered list
+     */
```
JavaDocs are all done well and nicely :)

wxliong reviewed (commented):
Only had one minor suggestion :)
Overall, codes are nicely formatted and understandable.

zhongfu committed:
Add Storage.save() that uses varargs

zhongfu committed:
Add find command to help message

zhongfu committed:
Refactor Duke to decouple chatbot logic and UI

Duke logic is tightly coupled to UI and makes assumptions about the
implementation details of the UI class (e.g. that it behaves like a
terminal, or that it can block and wait for user input).

This makes implementing other types of user interfaces (e.g. desktop
GUIs or web UIs) more cumbersome than necessary, such as displaying
different types of messages in different styles.

Representing the input/output to/from the chatbot logic as abstract
messages allows us to accomodate a larger variety of behaviors and
quirks for different UIs, and possibly easily expand the types of
messages that can be displayed in different styles in the future.

Refactoring chatbot logic out into a discrete class also allows us to
add support for different types of UIs with relative ease, and will
facilitate easier testing (as we can test the chatbot logic separately
from the UI logic).

zhongfu committed:
Implement rudimentary GUI for Duke

zhongfu committed:
Remove obsolete TUI test files

zhongfu committed:
Merge branch 'branch-Level-10'

zhongfu committed:
Add some assertions to validate application state

zhongfu committed:
Remove unnecessary imports

zhongfu committed:
Decrease scope for methods and constructors

Some of them have an unnecessarily wide scope, so we're reducing them

zhongfu committed:
Add and update javadoc comments; minor style fixes

Some newly refactored and/or added methods and
constructors are missing javadoc comments.

Style fixes are minor, e.g. extra lines removed, undescriptive
names changed

zhongfu committed:
Fix javadocs style issues

- subject-verb agreement (print -> prints)
- trailing whitespace

zhongfu committed:
Update ParserTest to avoid testing private methods

ParserTest tests parseArgString(), which was made private recently.

We can still test it indirectly through parseInput() (which also makes
more sense, since parseInput() is the bit that'll actually be used by
methods outside of Parser)

zhongfu committed:
Reduce complexity of Duke handleInput() method

It's kind of too complicated, but I'm also a little lazy, so I'm just
gonna break it up into multiple methods for each command.

They all have similar interfaces though, so if we do end up wanting to
make it Better^tm, then it shouldn't be too difficult.

zhongfu committed:
Add shorter aliases for todo/event/deadline

zhongfu committed:
Merge pull request #3 from zhongfu/branch-A-CodeQuality

Code quality improvements

zhongfu committed:
Merge pull request #2 from zhongfu/branch-A-Assertions

Assertions

zhongfu committed:
Merge branch 'branch-C-FriendlierSyntax'

zhongfu committed:
Remove trailing whitespace

Those don't conform to the Java Code Conventions.

zhongfu committed:
Wrap words in message bubbles

Some messages can be long enough that the resulting message bubble
exceeds the width of the window.

An obvious fix would have been to enable word wrapping. However, the
old MessageBubble implementation did not have the ability to resize
dynamically, as the sizes were determined on creation (based on the
bounds of the text).

We reimplement MessageBubble to allow them to resize dynamically,
which also allows us to wrap (and... unwrap?) words easily.

zhongfu committed:
Add min width for bubbles

...so that shorter messages don't look too weird

zhongfu committed:
Add title, enforce min height/width

zhongfu committed:
Wrap words in "system" messages too

zhongfu committed:
Merge branch 'branch-A-Varargs'

zhongfu committed:
Use Serializable instead of Object for Storage.save()

We can only take in instances that implement Serializable, so specifying
Object instead doesn't make much sense

zhongfu committed:
Add user guide and screenshot

zhongfu committed:
Update user guide

zhongfu committed:
Add command input history

zhongfu committed:
Make StorageTest work again

I think I broke it in d3e006b.

zhongfu committed:
Bump version to 0.2

zhongfu committed:
Ignore command submission if input field is empty

zhongfu committed:
Submit command only if text field is in focus

zhongfu committed:
Fix some code styling issues
